src/lib/BLE/devBLETelemetry.cpp

```diff
@@ -0,0 +1,301 @@
+    memset(&outBuffer[CRSF_TELEMETRY_TYPE_INDEX + 1], 0, LinkStatisticsFrameLength);
+
+    crsfPayloadLinkstatistics_s* pStats = (crsfPayloadLinkstatistics_s*)(&outBuffer[CRSF_TELEMETRY_TYPE_INDEX + 1]);
+    pStats->uplink_RSSI_1 = 120;
+    pStats->uplink_RSSI_2 = 120;
+    pStats->downlink_RSSI = 120;
+    pStats->uplink_SNR = -20;
+    pStats->downlink_SNR = -20;
+    pStats->rf_Mode = CRSF::LinkStatistics.rf_Mode;
+    pStats->uplink_TX_Power = CRSF::LinkStatistics.uplink_TX_Power;
+
+    BluetoothTelemetrySendLinkStatsPacketEx(outBuffer);
+}
+
+
+static void BluetoothTelemetrySendLinkStatsPacket()
+{
+    if (!CRSF::CRSFstate)
+    {
+        return;
+    }
+
+    uint8_t outBuffer[LinkStatisticsFrameLength + 4];
+    memcpy(&outBuffer[CRSF_TELEMETRY_TYPE_INDEX + 1], (byte *)&CRSF::LinkStatistics, LinkStatisticsFrameLength);
+    //CRSF::LinkStatistics.uplink_RSSI_1 is negative number
+    //we send negative numbers to handset. OpenTX handles it as signed byte.
+    //Over BLE we send positive numbers, like inav/betaflight expects (unsigned byte, positive value = -dbm)
+    crsfPayloadLinkstatistics_s* pStats = (crsfPayloadLinkstatistics_s*)(&outBuffer[CRSF_TELEMETRY_TYPE_INDEX + 1]);
+    pStats->uplink_RSSI_1 = -pStats->uplink_RSSI_1;
+    pStats->uplink_RSSI_2 = -pStats->uplink_RSSI_2;
+    pStats->downlink_RSSI = -pStats->downlink_RSSI;
+    BluetoothTelemetrySendLinkStatsPacketEx(outBuffer);
+}
+
+static void BluetoothTelemetrySendRCFrame()
+{
+    if (!CRSF::CRSFstate)
+    {
+        return;
+    }
+
+    uint8_t outBuffer[RCframeLength + 4];
+
+    rcPacket_t* p = (rcPacket_t*)(&outBuffer[0]);
+
+    p->header.device_addr = CRSF_ADDRESS_RADIO_TRANSMITTER;
+    p->header.frame_size = RCframeLength + 2;
+    p->header.type = CRSF_FRAMETYPE_RC_CHANNELS_PACKED;
+
+    p->channels.ch0 = ChannelData[0];
+    p->channels.ch1 = ChannelData[1];
+    p->channels.ch2 = ChannelData[2];
+    p->channels.ch3 = ChannelData[3];
+    p->channels.ch4 = ChannelData[4];
+    p->channels.ch5 = ChannelData[5];
+    p->channels.ch6 = ChannelData[6];
+    p->channels.ch7 = ChannelData[7];
+    p->channels.ch8 = ChannelData[8];
+    p->channels.ch9 = ChannelData[9];
+    p->channels.ch10 = ChannelData[10];
+    p->channels.ch11 = ChannelData[11];
+    p->channels.ch12 = ChannelData[12];
+    p->channels.ch13 = ChannelData[13];
+    p->channels.ch14 = ChannelData[14];
+    p->channels.ch15 = ChannelData[15];
+
+    outBuffer[CRSF_TELEMETRY_TYPE_INDEX + 1 + RCframeLength] = crsf_crc.calc((byte *)&outBuffer[CRSF_TELEMETRY_TYPE_INDEX], RCframeLength + 1 );
+
+    rcCRSF->setValue(outBuffer, RCframeLength + 4);
+    rcCRSF->notify();
+}
+
+void BluetoothTelemetryShutdown()
+{
+    if ( pServer != nullptr)
+    {
+        DBGLN("Stopping BLE Telemetry!");
+
+        NimBLEDevice::stopAdvertising();
+        NimBLEDevice::deinit(true);
+        pServer = nullptr;
+        rcCRSF = nullptr;
+        vSemaphoreDelete(_mutex);
+    }
+}
+
```

Comment:
```suggestion

static void assignRandomAddress()
 {
    uint8_t blead[6];
    memcpy( blead, NimBLEDevice::getAddress().getNative(), 6 );
    blead[5] |= 0xC0;  //random address should have 11 at most significat bits
    blead[0]++;
    NimBLEDevice::setOwnAddrType(BLE_OWN_ADDR_RANDOM);
    ble_hs_id_set_rnd(blead);
 }

 static void onAdvertisingStopped(NimBLEAdvertising *pAdv)
 {
    if (pServer != nullptr)
    {
        assignRandomAddress();
        pAdv->start(0, &onAdvertisingStopped);
    }
 }
 
```

Add address assignment method and advertisement callback

Reply:
that code doesn't work. Still a new address generated every time. So old code is used